### PR TITLE
BLD: Build shared c dependencies as a library

### LINKED ISSUE #30873: Parallel build_ext might fail due to file collisions
#### Code Sample, a copy-pastable example if possible

```bash
python setup.py build_ext --inplace -j 8
```
#### Problem description

See some discussion in #30862 .  I have 8 cores on Windows.  When I use `-j 8`, in the middle of the build, I will get this error message:

```
c:\Code\pandas_dev\pandas\pandas\_libs\tslibs\src\datetime\np_datetime.c : fatal error C1083: Cannot open compiler generated file: 'c:\Code\pandas_dev\pandas\build\temp.win-amd64-3.7\Release\pandas\_libs\tslibs\src\datetime\np_datetime.obj': Permission denied
```

What I think is happening is that in `setup.py`, `_libs.tslibs.conversion`, `libs.tslibs.np_datetime` and `_libs.tslibs.period` all have to compile `pandas/_libs/tslibs/src/datetime/np_datetime.c` .  If the parallel build is timed in such a way that two of those extensions are being built at the same time, then they can conflict in writing `np_datetime.obj`.  A similar thing can happen with `_libs.lib`, `_libs.parsers`, and `_libs.tslibs.parsing` all having to compile `pandas/_libs/src/parser/tokenizer.c` .

If I do two `python setup.py build_ext --inplace -j 8` steps in a row, then everything is fine.  Also things are fine with `-j 4`.  But I can imagine that even with `-j 4`, you could end up with the same kind of issue, even in CI.

I don't know enough about how extensions are built, distutils and setuptools in order to avoid this potential collision during the build process.

#### Output of ``pd.show_versions()``

<details>

c:\Code\pandas_dev\pandas\pandas\core\index.py:29: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.
  FutureWarning,

INSTALLED VERSIONS
------------------
commit           : 13858f6e11790bb0f52092d4f8ac65c4057d9e74
python           : 3.7.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.26.0.dev0+1734.g13858f6e1
numpy            : 1.17.3
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 42.0.2.post20191201
Cython           : 0.29.14
pytest           : 5.3.2
hypothesis       : 4.57.1
sphinx           : 2.3.1
blosc            : None
feather          : None
xlsxwriter       : 1.2.7
lxml.etree       : 4.4.2
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.10.1
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : 1.3.1
fastparquet      : 0.3.2
gcsfs            : None
lxml.etree       : 4.4.2
matplotlib       : 3.1.2
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.1
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
pytest           : 5.3.2
s3fs             : 0.4.0
scipy            : 1.3.1
sqlalchemy       : 1.3.12
tables           : 3.6.1
tabulate         : 0.8.6
xarray           : 0.14.1
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.7
numba            : 0.46.0
</details>

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #30873 (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Does anyone else want to test this? I can only test with -j4 on my old 2-core i3.

### CODE REVIEW

File: .github/actions/build_pandas/action.yml
Comment (mroeschke) [inline .github/actions/build_pandas/action.yml:19]: Might still be worth aligning this with the number of core on the GH hosted runners (2 Window/Linux, 3 MacOS): https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources=

**Comment by jbrockmendel** [comment]
Does this affect build size or speed or inlining or ...?

File: setup.py
Comment (jbrockmendel) [inline setup.py:657]: delete commented-out?

**Comment by Dr-Irv** [comment]
I like this PR, and want to try it out, but my laptop where I do 8-core builds on Windows that always fails is out of commission for a week or so.  Once I get that back, I will give it a try and let you know the outcome.

File: setup.py
Comment (WillAyd) [inline setup.py:454]: Do we need these `get_path("include")` calls? I think the compiler should handle that natively

File: setup.py
Comment (WillAyd) [inline setup.py:444]: Nit but maybe change this name to internal_libraries? My first thought when seeing ext_libraries was something installed on the system

**Comment by WillAyd** [comment]
@lithomas1 do you know what makes the compilation process different for these versus shared libraries created from compiled Cython modules? I'm slightly -1 on adding this complexity and certainly expanding it beyond these libraries for what seems like a very marginal gain

**Comment by Dr-Irv** [comment]
I just tried this PR to build the latest version on my Windows laptop using `python setup.py build_ext -j 8` and it worked perfectly.  I usually have to do it 2 or 3 times because of the conflicts.  So, as the person who created the original issue, I'm +8 (to match my cores) on this!

**Comment by lithomas1** [comment]
> @lithomas1 do you know what makes the compilation process different for these versus shared libraries created from compiled Cython modules? I'm slightly -1 on adding this complexity and certainly -1 for expanding it beyond these libraries for what seems like a very marginal gain

The problem is that there seems to be no way to link against other Extensions(the shared library made from the Cython module) from another Extension. The first option from [SO Link](https://stackoverflow.com/questions/57609741/link-against-another-python-c-extension#comment115546116_57609741) is what I'm doing (the other options don't seem to be too relevant, and are also going over my head). 

The other benefit of this is that the parser and the tslibs c sources don't get recompiled many times.

**Comment by mroeschke** [comment]
Would this potentially help solve https://github.com/pandas-dev/pandas/issues/47305?

**Comment by WillAyd** [comment]
> The problem is that there seems to be no way to link against other Extensions(the shared library made from the Cython module) from another Extension. The first option from [SO Link](https://stackoverflow.com/questions/57609741/link-against-another-python-c-extension#comment115546116_57609741) is what I'm doing (the other options don't seem to be too relevant, and are also going over my head).

Does the `depends` keyword argument not help here?

**Comment by simonjayhawkins** [comment]
> Would this potentially help solve #47305?

the longer #47305 remains unresolved, I think the higher chance of failure when bisecting regressions (but can maybe alter the bisect process if this becomes an issue)

so +1 here if solves #47305

**Comment by jreback** [comment]
ok let's merge this and see

**Comment by lithomas1** [comment]
> > The problem is that there seems to be no way to link against other Extensions(the shared library made from the Cython module) from another Extension. The first option from [SO Link](https://stackoverflow.com/questions/57609741/link-against-another-python-c-extension#comment115546116_57609741) is what I'm doing (the other options don't seem to be too relevant, and are also going over my head).
> 
> Does the `depends` keyword argument not help here?

I'm pretty sure I tried this and it didn't work, let me get back to you here.

**Comment by WillAyd** [comment]
Also fine with this for now if it fixes immediate issues. Can always clean up later. Also exploring different build tooling in #47380 which could help

**Review by WillAyd — APPROVED** [review]

File: .github/actions/build_pandas/action.yml
Comment (mroeschke) [inline .github/actions/build_pandas/action.yml:19]: Unless this address's @simonjayhawkins https://github.com/pandas-dev/pandas/pull/47081#issuecomment-1157504061= I would prefer to keep this at `N_JOBS: 1`

Otherwise, I would suggest this be specific to the number of cores available on GHA hosted runners by OS: https://github.com/pandas-dev/pandas/pull/47081#discussion_r878982485=

**Review by mroeschke — CHANGES_REQUESTED** [review]

File: .github/actions/build_pandas/action.yml
Comment (simonjayhawkins) [reply .github/actions/build_pandas/action.yml:19]: > Unless this address's @simonjayhawkins [#47081 (comment)](https://github.com/pandas-dev/pandas/pull/47081#issuecomment-1157504061)

I've needed to restart a few workflows this last week or so as they fail on the initial build, but it maybe once it is fixed that the initial build is successful, that subsequent incremental builds for commits in the range of the last couple of weeks could also be successful.

And it could also be that if this is a reoccurring problem that I can change the workflow to repeat the previous build step if the pandas import fails.

so I am happy that any comments/concerns are fully addressed before merging this.

**Comment by Dr-Irv** [comment]
@lithomas1 see my suggested changes in https://github.com/lithomas1/pandas/pull/16

**Comment by lithomas1** [comment]
I haven't touched this PR in a while, since we are considering Meson(working on exploring this front)/Cmake as our new build system. If the other core devs still think this is worthwhile, I am happy to pick it back up though.

**Comment by lithomas1** [comment]
It might be time to revive this. This might fix the CI and I don't think the meson work will be ready before at least the end of this year. cc @WillAyd

**Comment by WillAyd** [comment]
Hmm I _think_ I'd prefer not to hack setuptools like this versus just using Meson. Stuck with anything in particular on that I might be able to help with?

**Comment by Dr-Irv** [comment]
> FWIW if you look at the build log this still generates the `np_datetime.o` file multiple times. Assuming a race condition between the various compilation / linking phases is the root of the issue I'm not sure this solves anything

See my message above:  https://github.com/pandas-dev/pandas/pull/47081#issuecomment-1157912923

I created a PR for @lithomas1 that fixes this issue.

**Comment by WillAyd** [comment]
Hmm interesting. It looks like the build log shows that file still getting generated a few times. 

Not opposed to merging this as is just don't think this does what we think it does.  Just want to be wary of continued setuptools hacks going forward

**Comment by lithomas1** [comment]
> Hmm I _think_ I'd prefer not to hack setuptools like this versus just using Meson. Stuck with anything in particular on that I might be able to help with?

I working out a couple of bugs in meson and meson-python. mesonbuild/meson#10639 is the one I'm looking into right now.


> > FWIW if you look at the build log this still generates the `np_datetime.o` file multiple times. Assuming a race condition between the various compilation / linking phases is the root of the issue I'm not sure this solves anything
> 
> See my message above: [#47081 (comment)](https://github.com/pandas-dev/pandas/pull/47081#issuecomment-1157912923)
> 
> I created a PR for @lithomas1 that fixes this issue.

Merged it.



> Hmm interesting. It looks like the build log shows that file still getting generated a few times.
> 
> Not opposed to merging this as is just don't think this does what we think it does. Just want to be wary of continued setuptools hacks going forward

Yeah, this could end up not worth it ifit causes a performance egression somewhere or somehow borks the wheels.

**Comment by rhshadrach** [comment]
Around 50% of the time I have to rebuild due to the undefined symbol issue. I have a 16 core (32 threads) machine and with this PR running

```
find . -type f -exec touch {} +
python setup.py build_ext -j 32
```

three times all builds were successful. Doing the same on main, all three failed. Based on the conversation above, I think that makes me +32 on this PR.

**Comment by lithomas1** [comment]
Small update on this:

Thanks all for the encouragement on this, and sorry for the long period without updates. 
I think I planned to get this in in the meantime before I finished meson work, but it isn't worth it at this point to spend the time investigating perf, and other issues that this may cause.

Work on meson is a mixed bag

- I can now compile and test on CI on Windows, macOS, and Linux.
- Sadly, we have failing tests. :(
        - The only issue really worth worrying about is the SAS failures. Looks like there is a problem exporting the correct module name there for the Cython Extension Module.
        - I think some of our tests also depend on pandas being built inplace so which is causing some of the failures.
- Finally, I have to figure out how to generate the pandas version from the tag. This will fix most of the other failures.

The bad news is I still don't think we'll be able to get meson enabled as the default build system before the end of the year. I've been upstreaming fixes to both the meson and meson-python libraries, and probably is going to be a while before they cut another release. 

The good news is that we'll probably have some sort of meson support in main by next month(hopefully?) for brave people to try, it just won't be in the CI until meson and meson-python cut a new release with my fixes in them.

**Comment by WillAyd** [comment]
> Finally, I have to figure out how to generate the pandas version from the tag. This will fix most of the other failures.

This is managed today via versioneer, which I think is only supported via setuptools. Might need to look at upstreaming support for meson in that as well

**Comment by rgommers** [comment]
Re the version, it depends on what you want:
1. Have an internal version string which contains the commit hash, but a project version (ending up in the wheel filename) of `1.6.0.dev0` without git hash.
2. Also have the git hash reflected in the wheel version. This is what a number of projects using `versioneer` do.
3. Getting the project version directly from a git tag, ala `setuptools_scm`.

(2) isn't possible easily in Meson right now, because using its `vcs_tag` feature in the project version would trigger a full reconfigure -see https://github.com/mesonbuild/meson/issues/688#issuecomment-422757933. I'm not sure it would be accepted, although there is recent movement on the issue I linked (including a workaround implementation).

(3) seems to be more feasible.

SciPy used to do (2), and I switched it to (1) during the move to Meson. That has been fine so far. The benefit of having the git hash in the version is relatively minor, and it's still available at install time as `python -c "import scipy; print(scipy.__version__)"`.

**Comment by lithomas1** [comment]
> SciPy used to do (2), and I switched it to (1) during the move to Meson. That has been fine so far. The benefit of having the git hash in the version is relatively minor, and it's still available at install time as `python -c "import scipy; print(scipy.__version__)"`.

I'm a little worried that this will overwrite nightly wheels, though, since the filenames will be the same. This could be a problem if someone downstream is pinning the tests of pandas nightlies.

Is there something we can do in meson-python to workaround this? Since the version in meson has to be hard-coded, it might make sense for meson-python to take an option in the pyproject.toml, to get the version from git?

**Comment by rgommers** [comment]
> Is there something we can do in meson-python to workaround this? Since the version in meson has to be hard-coded, it might make sense for meson-python to take an option in the pyproject.toml, to get the version from git?

That sounds like a reasonable thing to do in `meson-python`. I think all that is needed is to (a) actually retrieve the git-based version, and (b) use that in the wheel name and in the `METADATA` file. That said:
- this stuff is quite fragile (`versioneer` seems to be a never-ending source of bugs, it's way over-complicated), and
- different projects may want to make different choices regarding what the version numbering should be (e.g., just the git hash vs. an incrementing count since the last tag - the latter can be useful for nigtlies but problematic with shallow clones)

So perhaps rather than baking it into `meson-python` itself, there could be some hook mechanism so projects can provide their own function which returns the desired version string. WDYT?

**Comment by WillAyd** [comment]
Will have to excuse any ignorance in the question but is it possible to continue to use setuptools with meson?

**Comment by rgommers** [comment]
It's perfectly fine to have two build systems live side-by-side in this repo. SciPy has had both `setuptools` and `meson` since last Christmas. The one choice to make is which of the two build systems gets invoked for `pip install .` - there's a single hook in `pyproject.toml`, so you cannot make that configurable.

**Comment by WillAyd** [comment]
If that's the case I would think it's probably easier to just stick with setuptools for the initial cutover then change to meson-Python at a later date. At the very least that way can move forward without having to solve the versioning issue

**Comment by WillAyd** [comment]
Actually re-reading your comment I think we are talking about different things, but moved reply to https://github.com/lithomas1/pandas/pull/19#issuecomment-1258820994

**Pull request closed without merging**
